File: src/main/java/no/skatteetaten/aurora/databasehotel/dao/DatabaseManager.java
package no.skatteetaten.aurora.databasehotel.dao;

import java.util.List;
import java.util.Map;
import java.util.Optional;

import javax.sql.DataSource;

import no.skatteetaten.aurora.databasehotel.dao.dto.Schema;

public interface DatabaseManager {
    boolean schemaExists(String schemaName);

    String createSchema(String schemaName, String password);

    void updatePassword(String schemaName, String password);

    Optional<Schema> findSchemaByName(String name);

    List<Schema> findAllNonSystemSchemas();

    void deleteSchema(String schemaName);

    void executeStatements(String... statements);

    String getCurrentUserName();

    List<Map<String, Object>> query(String query, Object... params);

    DataSource getDataSource();
}

File: src/main/kotlin/no/skatteetaten/aurora/databasehotel/dao/DatabaseManager.kt
package no.skatteetaten.aurora.databasehotel.dao

import java.util.Date

data class Schema @JvmOverloads constructor(
    val username: String,
    val lastLogin: Date? = null
)

interface DatabaseManager {

    fun schemaExists(schemaName: String): Boolean

    fun createSchema(schemaName: String, password: String): String

    fun updatePassword(schemaName: String, password: String)

    fun findSchemaByName(schemaName: String): Schema?

    fun findAllNonSystemSchemas(): List<Schema>

    fun deleteSchema(schemaName: String)

    fun executeStatements(vararg statements: String)

    fun getMaxTablespaces(): Int?

    fun getUsedTablespaces(): Int?
}
